### PR TITLE
OCPBUGS-31067: Keep /var/lib/kubelet/kubeconfig

### PR DESCRIPTION
**Pull request by cdoern**
Currently we delete and re-generate the entire KC when rotating api-int. Keep the file and just change the cert authority data

We still need to restart kubelet.

### CODE REVIEW

File: templates/common/gcp/files/etc-kubernetes-credential-providers-gcr-credential-provider.yaml [hunk 1]
```diff
@@ -6,9 +6,13 @@ contents:
     kind: CredentialProviderConfig
     providers:
       - name: gcr-credential-provider
+        apiVersion: credentialprovider.kubelet.k8s.io/v1
```
Comment (yuqi-zhang) [inline templates/common/gcp/files/etc-kubernetes-credential-providers-gcr-credential-provider.yaml:9]: Hmm, what are these changes for?

File: pkg/daemon/certificate_writer.go [hunk 2]
```diff
@@ -340,41 +339,40 @@ func (dn *Daemon) syncControllerConfigHandler(key string) error {
...
 			} else if err != nil {
-				return false, fmt.Errorf("unexpected error reading kubeconfig file, %v", err)
+				return fmt.Errorf("unexpected error reading kubeconfig file, %v", err)
 			}
-			currentKC := clientcmdv1.Config{}
-			err = yaml.Unmarshal(f, &currentKC)
+			kubeletKC := clientcmdv1.Config{}
+			err = yaml.Unmarshal(f, &kubeletKC)
+			// set CA data to the one we just parsed above, the rest of the data should be preserved.
+			kubeletKC.Clusters[0].Cluster.CertificateAuthorityData = currentKC.Clusters[0].Cluster.CertificateAuthorityData
```
Comment (yuqi-zhang) [inline pkg/daemon/certificate_writer.go:357]: This should be fine but I wonder if `Clusters` could ever contain more than 1 entry

File: pkg/daemon/certificate_writer.go [hunk 3]
```diff
@@ -340,41 +339,40 @@ func (dn *Daemon) syncControllerConfigHandler(key string) error {
...
-		}
-		if err := runCmdSync("systemctl", "daemon-reload"); err != nil {
-			return err
-		}
-
-		if err := runCmdSync("systemctl", "restart", "kubelet"); err != nil {
-			return err
-		}
-
-		if err := wait.PollUntilContextTimeout(context.TODO(), 10*time.Second, 2*time.Minute, false, func(ctx context.Context) (bool, error) {
```
Comment (yuqi-zhang) [inline pkg/daemon/certificate_writer.go:359]: This poll used to be here for us to wait until `/var/lib/kubelet/kubeconfig` get regenerated, right?

**Review by yuqi-zhang — COMMENTED** [review]
Generally lgtm logic wise. A few questions/comments inline

File: pkg/daemon/certificate_writer.go [hunk 4]
```diff
@@ -340,41 +339,40 @@ func (dn *Daemon) syncControllerConfigHandler(key string) error {
...
-				return false, nil
+				return err
 			} else if err != nil {
-				return false, fmt.Errorf("unexpected error reading kubeconfig file, %v", err)
+				return fmt.Errorf("unexpected error reading kubeconfig file, %v", err)
 			}
-			currentKC := clientcmdv1.Config{}
-			err = yaml.Unmarshal(f, &currentKC)
+			kubeletKC := clientcmdv1.Config{}
+			err = yaml.Unmarshal(f, &kubeletKC)
```
Comment (yuqi-zhang) [inline pkg/daemon/certificate_writer.go:355]: This err isn't actually being handled

File: pkg/daemon/certificate_writer.go (hunk 2 above)
Comment (cdoern) [reply pkg/daemon/certificate_writer.go:357]: I never have seen that, clusters[0] is always the data we are looking for.

File: templates/common/gcp/files/etc-kubernetes-credential-providers-gcr-credential-provider.yaml (hunk 1 above)
Comment (cdoern) [reply templates/common/gcp/files/etc-kubernetes-credential-providers-gcr-credential-provider.yaml:9]: weird, I didn't add that. Must have snuck in, will remove.

File: pkg/daemon/certificate_writer.go (hunk 3 above)
Comment (cdoern) [reply pkg/daemon/certificate_writer.go:359]: yeah, which we don't need since we aren't deleting it

File: pkg/daemon/certificate_writer.go (hunk 4 above)
Comment (cdoern) [reply pkg/daemon/certificate_writer.go:355]: thanks, will handle this

File: templates/common/gcp/files/etc-kubernetes-credential-providers-gcr-credential-provider.yaml [hunk 5]
```diff
@@ -12,3 +12,7 @@ contents:
           - "container.cloud.google.com"
         defaultCacheDuration: "0s"
         apiVersion: credentialprovider.kubelet.k8s.io/v1
+        defaultCacheDuration: "1m"
+        args:
+          - get-credentials
+          - --v=3
```
Comment (JoelSpeed) [inline templates/common/gcp/files/etc-kubernetes-credential-providers-gcr-credential-provider.yaml:18]: Unrelated changes from a different PR? Huh? Needs a rebase?

Comment (cdoern) [reply templates/common/gcp/files/etc-kubernetes-credential-providers-gcr-credential-provider.yaml]: sorry, yeah. Fixed now.

**Comment by yuqi-zhang** [comment]
/close

Merged as part of https://github.com/openshift/machine-config-operator/pull/4276
